src/storage-client/src/controller/hosts.rs

```diff
@@ -337,3 +357,57 @@ where
         Ok(())
     }
 }
+
+/// A helper that shares state with a [`StorageHosts`] and allows repeatedly
+/// fetching metrics for all known storage hosts.
+pub struct MetricsFetcher {
+    /// The orchestrator that we use to fetch metrics.
+    orchestrator: Arc<dyn NamespacedOrchestrator>,
+    /// The shared assignment of storage objects to storage hosts. This is
+    /// shared with a [`StorageHosts`].
+    objects: Arc<Mutex<HashMap<GlobalId, StorageHostAddr>>>,
+}
+
+impl MetricsFetcher {
+    /// Creates a new [`MetricsFetcher`].
+    fn new(
+        orchestrator: Arc<dyn NamespacedOrchestrator>,
+        objects: Arc<Mutex<HashMap<GlobalId, StorageHostAddr>>>,
+    ) -> Self {
+        Self {
+            orchestrator,
+            objects,
+        }
+    }
+
+    /// Fetches and returns metrics for all known storage hosts.
+    pub async fn fetch_metrics(
+        &mut self,
+    ) -> Vec<(
+        GlobalId,
+        String,
+        Result<Vec<ServiceProcessMetrics>, anyhow::Error>,
+    )> {
+        let service_ids = {
+            self.objects
+                .lock()
+                .await
+                .iter()
+                .map(|(id, host_addr)| (id.clone(), host_addr.clone()))
+                .collect::<Vec<_>>()
+        };
+        tracing::trace!("fetch_metrics: service_ids: {:?}", service_ids);
+
+        let mut metrics = Vec::new();
+
+        for (service_id, host_addr) in service_ids {
+            let service_metrics = self
+                .orchestrator
+                .fetch_service_metrics(&service_id.to_string())
+                .await;
+            metrics.push((service_id, host_addr, service_metrics));
```

Comment:
this is going to be aggressively slow when we have 1500 sources, can we add a TODO to run these concurrently?

Reply:
👍